### PR TITLE
Return zero array from np.dot when the arguments are empty.

### LINKED ISSUE #5539: Invalid np.dot() output for vector/matrix multiply of empty arrays
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Looks like `np.dot` implementation doesn't handle empty input properly:

```python
In [161]: X = np.array([]).reshape(0, 2)

In [162]: Y = np.array([])

In [163]: numba.njit(lambda X, Y: np.dot(Y, X))(X, Y)
Out[163]: array([-1.36311572e+57, -1.36311572e+57])

In [164]: np.dot(Y, X)
Out[164]: array([0., 0.])
```

### PR DESCRIPTION
**Pull request by AndrewEckart**
When np.dot(a, b) is called with empty argments, the result can be non-empty if the zero dimension is eliminated (e.g. a is m x 0 and b is 0 x n). In these cases, we should return an array of zeros. Previously, the result was initialized with np.empty, so the it contained garbage values.

Fixes #5539

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch. The fix seems good. Please could you add/update appropriate unit tests in here: https://github.com/numba/numba/blob/master/numba/tests/test_linalg.py#L30 to ensure that the change is correct. Many thanks.

**Comment by AndrewEckart** [comment]
Seems my fix is working for `dot_2`, but `dot_3` also needs to be fixed, and I'm not quite familiar enough with the lowered API to get it working. I was trying to do something along the lines of the `numpy_zeros_nd` implementation with `_zero_fill_array`.

**Comment by AndrewEckart** [comment]
Okay, I got it working for both `dot_2` and `dot_3`. I'm very much a beginner with the IR Builder, so please let me know if there's a cleaner way to write things.

File: numba/np/linalg.py [hunk 1]
```diff
@@ -611,8 +618,15 @@ def dot_3_vm(context, builder, sig, args):
     for val in m_shapes:
         check_c_int(context, builder, val)
 
-    call_xxgemv(context, builder, do_trans, mty, m_shapes, m_data,
-                v_data, out.data)
+    zero = ir.Constant(intp_t, 0)
```
Comment (stuartarchibald) [inline numba/np/linalg.py:621]: There's a preference in the code base for using the `context` where possible, e.g.
`zero = context.get_constant(types.intp, 0)`
(same applies throughout).

File: numba/tests/test_linalg.py [hunk 2]
```diff
@@ -169,16 +168,21 @@ def samples(m, n):
         cfunc2 = jit(nopython=True)(pyfunc2)
         if pyfunc3 is not None:
             cfunc3 = jit(nopython=True)(pyfunc3)
-        for a, b in samples(m, n):
-            self.check_func(pyfunc2, cfunc2, (a, b))
-            self.check_func(pyfunc2, cfunc2, (b, a.T))
-        if pyfunc3 is not None:
+
+        for m, n in [(2, 3),
+                     (3, 0)
```
Comment (stuartarchibald) [inline numba/tests/test_linalg.py:173]: What about e.g. `(0, 3)` ?

File: numba/tests/test_linalg.py [hunk 3]
```diff
@@ -229,10 +233,12 @@ def samples(m, n, k):
 
         # Test generic matrix * matrix as well as "degenerate" cases
         # where one of the outer dimensions is 1 (i.e. really represents
-        # a vector, which may select a different implementation)
+        # a vector, which may select a different implementation),
+        # or both matrices are empty but their product is not.
         for m, n, k in [(2, 3, 4),  # Generic matrix * matrix
                         (1, 3, 4),  # 2d vector * matrix
                         (1, 1, 4),  # 2d vector * 2d vector
+                        (3, 2, 0)   # Empty matrix * empty matrix
```
Comment (stuartarchibald) [inline numba/tests/test_linalg.py:241]: What about e.g. `(0, 0, 3)` ?

**Comment by AndrewEckart** [comment]
> With the new use cases, I think they will probably need explicitly catching as per the current "inner dimension is zero" cases, at present, these new cases are manifesting as an error in the BLAS call, `XERBLA` is complaining!

Updated to deal with cases where one or more arguments is empty because an outer dimension is zero and added test cases to cover each case.

It's a little weird because in these cases, the output is empty, but I'm still calling `cgutils.memset` with `size=0` (which does nothing). I could add more conditional logic to avoid that if it would be preferable.

**Comment by stuartarchibald** [comment]
> > With the new use cases, I think they will probably need explicitly catching as per the current "inner dimension is zero" cases, at present, these new cases are manifesting as an error in the BLAS call, `XERBLA` is complaining!
> 
> Updated to deal with cases where one or more arguments is empty because an outer dimension is zero and added test cases to cover each case.
> 
> It's a little weird because in these cases, the output is empty, but I'm still calling `cgutils.memset` with `size=0` (which does nothing). I could add more conditional logic to avoid that if it would be preferable.

Thanks for patching that too. Until demonstrated as a bottleneck, I think just leaving a `memset` of zero bytes is fine, it's a function call vs branch, and then obviously complexity that goes into handling it.

**Pull request merged**